# PseudoFuN
This Shiny app is made by Travis S Johnson and Zhi Huang. The pseudogene alignment database
is generated by Travis S Johnson and Sihong Li using the Ohio Supercomputing Center. A
second query tool which allows for direct sequence queries is also available through the Ohio
Supercomputing Center developed by Eric Franz. The online Shiny app is still under development.
The app has been tested on OSX El Captitan 10.11.6 and R version 3.4.2. The website version
has been tested in Chrome, Firefox, and Safari.

## What does PseudoFuN do?
Pseudogene-gene families are sets of homologous genes which were assigned pseudogenes
based on similar sequence homology. This app displays a set of the more than 26000
pseudogene-gene families that were generated using our pipeline. This database allows a
user to query gene symbols, ensembl gene IDs and entrez gene IDs in our database. The
peudogene-gene family networks containing these queries are then displayed interactively
using networkD3. The genes in these networks can also be used to run Gene Ontology analysis
with topGO to assess the functional potentials of these pseudogenes. A more extensive query
tool is also developed in association with the Ohio Super Computing Center that allows
users to not only query gene names but also query sequences which are then included into
their assigned network. This application also allows users to interactively view the GO
information for genes in the network individually, run multiple queries simultaniously and
download the results to a csv file.  For access to the more expansive tool please contact
Travis S Johnson (travis.johnson@osumc.edu) or Dr. Yan Zhang (yan.zhang@osumc.edu).

## Running PseudoFuN
### Network Analysis
After visiting our URL you must select the Database you wish to use and input a gene or
pseudogene to query. Once these are selected clicking on the tabs will display the networks
for pseudogene-gene families that contain the respective gene/pseudogene. Gene symbols,
Ensembl gene IDs (for genes), Ensembl transcript IDs (for pseudogenes), and Entrez IDs are
supported. The resulting graphs when nodes are selected will display gene/pseudogene, the
gene symbol if available, and the Ensembl ID.

### GO Analysis (Long runtime):
If the Run GO Analysis option is selected then all of the genes within the networks will be
aggregated into a single gene list. This gene list will then be used as input against all
other genes into the topGO R package. The resulting table displays GO terms with their
associated p-values and number of assigned genes. The GO term set must also be selected.
Options include Biological Process, Molecular Function or Cellular Component. The default
method is using a Fishers Exact test but other options include using Kolmogorov-Smirnov (KS
Classic), and Kolmogorov-Smirnov Elim (KS Elim, Alexa et. al. 2016). The table will always
be sorted by the Fishers Exact p-value. The GO terms that have no assigned genes are removed
unless the "Include GO terms without any assigned genes" option is selected.

### Genomic Mapping analysis
All of the genes that are contained in the networks identified by the query can be mapped
back to the genome. These graphs are meant to show if the genes/pseudogenes localize to
specific regions of the genome and are displayed as a circular plot.

This application is still being tested. For any bugs please contact (travis.johnson@osumc.edu).
